src/lib/logstore/log_dev.cpp

```diff
@@ -399,233 +399,152 @@ bool LogDev::flush_if_needed(int64_t threshold_size) {
     THIS_LOGDEV_LOG(TRACE, "Flush completed for logid[{} - {}]", lg->m_flush_log_idx_from, lg->m_flush_log_idx_upto);
 
     m_log_records->complete(lg->m_flush_log_idx_from, lg->m_flush_log_idx_upto);
     m_last_flush_idx = lg->m_flush_log_idx_upto;
-    const auto flush_ld_key = logdev_key{m_last_flush_idx, lg->m_log_dev_offset + lg->header()->total_size()};
     m_last_crc = lg->header()->cur_grp_crc;
 
     auto from_indx = lg->m_flush_log_idx_from;
     auto upto_indx = lg->m_flush_log_idx_upto;
     auto dev_offset = lg->m_log_dev_offset;
     for (auto idx = from_indx; idx <= upto_indx; ++idx) {
         auto& record = m_log_records->at(idx);
-        on_io_completion(record.store_id, logdev_key{idx, dev_offset}, flush_ld_key, upto_indx - idx, record.context);
+        logstore_req* req = s_cast< logstore_req* >(record.context);
+        HomeLogStore* log_store = req->log_store;
+        HS_LOG_ASSERT_EQ(log_store->get_store_id(), record.store_id,
+                         "Expecting store id in log store and flush completion to match");
+
+        log_store->on_write_completion(req, logdev_key{idx, dev_offset});
     }
-    lg->m_post_flush_process_done_time = Clock::now();
+    if (m_log_records_safe_truncate_to_idx < upto_indx) m_log_records_safe_truncate_to_idx = upto_indx;
 
-    HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_done_msg_time_ns,
-                      get_elapsed_time_us(lg->m_flush_finish_time, lg->m_post_flush_msg_rcvd_time));
+    HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_time_us,
+                      get_elapsed_time_us(m_last_flush_time, done_time));
     HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_post_flush_processing_latency,
-                      get_elapsed_time_us(lg->m_post_flush_msg_rcvd_time, lg->m_post_flush_process_done_time));
+                      get_elapsed_time_us(done_time));
     free_log_group(lg);
-    unlock_flush();
-}
 
-bool LogDev::run_under_flush_lock(const flush_blocked_callback& cb) {
-    {
-        std::unique_lock lk{m_block_flush_q_mutex};
-        if (m_stopped) {
-            THIS_LOGDEV_LOG(WARN, "Trying to lock a flush on a stopped logdev, not locking the flush");
-            return false;
-        }
-
-        bool expected_flushing{false};
-        if (!m_is_flushing.compare_exchange_strong(expected_flushing, true, std::memory_order_acq_rel)) {
-            // Flushing is blocked already, add it to the callback q
-            if (m_block_flush_q == nullptr) { m_block_flush_q = sisl::VectorPool< flush_blocked_callback >::alloc(); }
-            m_block_flush_q->emplace_back(cb);
-            return false;
-        }
+    // Truncate the log records, as it is no longer needed to be kept in memory. Everytime there is a read, we actually
+    // read from the drives.
+    // TODO:: we can keep this in memory to accelerate read until it is truncated
+    if (from_indx == m_log_records_last_truncate_to_idx + 1) {
```

Comment:
can we have a policy to keep last N records in memory?

Reply:
removed